File: TUM_Lens/app/src/main/java/com/maxjokel/lens/ModelSelectorFragment.java
package com.maxjokel.lens;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.ColorStateList;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.core.widget.CompoundButtonCompat;
import androidx.fragment.app.Fragment;
import android.view.HapticFeedbackConstants;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;


/*
 *  FRAGMENT FOR 'MODEL SELECTOR' ELEMENT IN BOTTOM SHEET
 *
 * -> we use a fragment here, because we reuse this component in
 *    'ViewFinder' as well as in
 *    'CameraRoll'
 *
 * -> additionally, the actual selector UI is created dynamically based on the nets.json file
 *    in the /assets directory that this fragment obtains via the ListSingleton instance
 *
 *
 * -> should be the most elegant solution to this
 *
 *
 * */

public class ModelSelectorFragment extends Fragment {

    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    // instantiate new SharedPreferences object

    SharedPreferences prefs = null;
    SharedPreferences.Editor prefEditor = null;


    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    // related to 'ListSingleton'

    ListSingleton listSingletonInstance = ListSingleton.getInstance();
    List<ModelConfig> MODEL_LIST = listSingletonInstance.getList();

    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    // related to 'NewStaticClassifier'

    NewStaticClassifier newStaticClassifier = NewStaticClassifier.getInstance();


    // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -


    // empty constructor
    public ModelSelectorFragment() {}

    public static ModelSelectorFragment newInstance() {
        return new ModelSelectorFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {

        // Please note the execution pipeline
        //   1. onCreate()
        //   2. onCreateView()   -> layout gets inflated
        //   3. onViewCreated()  -> we can now access the layout components
        //
        //
        // we dynamically create the RadioButtons in onViewCreated() for that reason


        // load sharedPreferences object and set up editor
        prefs = Objects.requireNonNull(this.getActivity()).getSharedPreferences("TUM_Lens_Prefs", Context.MODE_PRIVATE);

        super.onCreate(savedInstanceState);

    }

    @Override
    public View onCreateView(LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {

        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_model_selector, container, false);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

        // this is called right after 'onCreateView()'


    // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +
    // +          DYNAMICALLY SET UP RADIO BUTTONS             +
    // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +


        // workaround for 'dp' instead of 'px' units
        float dpRatio = view.getContext().getResources().getDisplayMetrics().density;

        // init RadioGroup
        RadioGroup radioGroup = view.findViewById(R.id.modelSelector_RadioGroup);


        // iterate over list of models and create RadioButtons
        for (int i = 0; i < MODEL_LIST.size(); i++){

            ModelConfig m = MODEL_LIST.get(i);

            // create new RadioButton
            RadioButton radioButton = new RadioButton(view.getContext());

            // set ModelConfig attributes
            radioButton.setId(m.getId());
            radioButton.setText(m.getName());
            radioButton.setTag(m.getModelFilename());

            // set appearance
            radioButton.setTextColor(ContextCompat.getColor(view.getContext(), R.color.black));
            radioButton.setTextSize(14.0f);
            LinearLayout.LayoutParams paramsButton = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
            ); // [source: https://stackoverflow.com/a/12728484]
            paramsButton.setMargins(0,(int)(8 * dpRatio),0,0); // dp
            radioButton.setLayoutParams(paramsButton);
            CompoundButtonCompat.setButtonTintList(radioButton,
                    ColorStateList.valueOf(ContextCompat.getColor(view.getContext(), R.color.colorPrimary)));

            // create the 'info' TextView
            TextView textView = new TextView(view.getContext());

            textView.setText("Top 5 accuracy: " + m.getTop5accuracy());

            // set appearance
            LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
            );
            params.setMargins((int)(33 * dpRatio),(int)(-4 * dpRatio),0,0); // dp
            textView.setLayoutParams(params);
            textView.setTextColor(ContextCompat.getColor(view.getContext(), R.color.light_grey));
            textView.setTextSize(12.0f);

            // add both elements to the RadioGroup
            radioGroup.addView(radioButton);
            radioGroup.addView(textView);

        } // END of FOR loop


    // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +
    // +           set INITIAL RadioGroup SELECTION            +
    // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +

        // get saved model Id from SharedPreferences
        int id = prefs.getInt("model", 0);
        RadioButton r = view.findViewById(id);

        if(id == 0 || r == null){
            r = view.findViewById(R.id.radioButton_FloatMobileNet);
        }

        // set initial selection
        r.setChecked(true);


    // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +
    // +           NETWORK SELECTOR via RadioGroup             +
    // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +

        // init SharedPreferences Editor
        prefEditor = prefs.edit();

        // init RadioGroup event listener
        radioGroup.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {

            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId) {

                // perform haptic feedback
                group.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_PRESS);

                int modelId;

                // check if it is the default RadioButton
                if(checkedId == R.id.radioButton_FloatMobileNet){
                    modelId = 0;
                } else {
                    modelId = view.findViewById(checkedId).getId();
                }

                // save selection to SharedPreferences
                prefEditor.putInt("model", modelId);
                prefEditor.apply();


                // trigger classifier update
                NewStaticClassifier.onConfigChanged();

            }
        });

    } // END of onViewCreated(...) - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

}

File: TUM_Lens/app/src/main/java/com/maxjokel/lens/fragments/ModelSelectorFragment.kt
package com.maxjokel.lens.fragments

import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.content.res.ColorStateList
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.os.Handler
import android.util.Log
import android.view.*
import android.widget.*
import androidx.core.content.ContextCompat
import androidx.core.widget.CompoundButtonCompat
import androidx.fragment.app.Fragment
import com.maxjokel.lens.R
import com.maxjokel.lens.classification.Classifier
import com.maxjokel.lens.classification.ListSingleton
import com.maxjokel.lens.helpers.App
import com.maxjokel.lens.helpers.DownloadFiles
import java.io.File

/* FRAGMENT FOR 'MODEL SELECTOR' ELEMENT IN BOTTOM SHEET
*  we use a fragment here, because we reuse this component in ViewFinder and CameraRoll
*  additionally, the actual selector UI is created dynamically based on the nets.json file
*  in the /assets directory that this fragment obtains via the ListSingleton instance
*/
class ModelSelectorFragment: Fragment() {

    private lateinit var prefs: SharedPreferences
    private lateinit var prefEditor: SharedPreferences.Editor

    private var MODEL_LIST = ListSingleton.modelConfigs

    private lateinit var downloadFiles : DownloadFiles

    override fun onCreate(savedInstanceState: Bundle?) {
        Classifier.initialize()
        // load sharedPreferences object and set up editor
        prefs = this.requireActivity()
            .getSharedPreferences("TUM_Lens_Prefs", Context.MODE_PRIVATE)
        prefEditor = prefs.edit()
        downloadFiles = DownloadFiles(this.requireContext())
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_model_selector, container, false)
    }

    @SuppressLint("SetTextI18n")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {

        val root : String = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            App.context!!.getExternalFilesDir(null)!!.path + "/models"
        }else{
            Environment.getExternalStorageDirectory().absolutePath + "/models"
        }

        val path = File(root)

        // workaround for 'dp' instead of 'px' units
        val dpRatio = view.context.resources.displayMetrics.density

        // init RadioGroup
        val radioGroup = view.findViewById<RadioGroup>(R.id.modelSelector_RadioGroup)

        // iterate over list of models and create RadioButtons
        for (m in MODEL_LIST) {
            // create new RadioButton
            val radioButton = RadioButton(view.context)

            // set ModelConfig attributes
            radioButton.id = m.modelId
            radioButton.text = m.modelName
            radioButton.tag = m.modelFilename

            // set appearance
            radioButton.setTextColor(ContextCompat.getColor(view.context, R.color.black))
            radioButton.textSize = 14.0f
            val paramsButton = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            ) // [source: https://stackoverflow.com/a/12728484]
            paramsButton.setMargins(0, (8 * dpRatio).toInt(), 0, 0) // dp
            radioButton.layoutParams = paramsButton
            CompoundButtonCompat.setButtonTintList(radioButton,
                ColorStateList.valueOf(ContextCompat.getColor(view.context, R.color.colorPrimary))
            )

            val exactPath = File("$path/${radioButton.tag}")

            radioButton.isClickable = exactPath.exists()


            // create the 'info' TextView
            val textView = TextView(view.context)
            textView.text = "" + m.top5accuracy

            // set appearance
            val params = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            )
            params.setMargins((33 * dpRatio).toInt(), (-4 * dpRatio).toInt(), 0, 0) // dp
            textView.layoutParams = params
            textView.setTextColor(ContextCompat.getColor(view.context, R.color.black))
            textView.textSize = 12.0f


            // add download button
            val button = Button(view.context)
            button.tag = m.modelFilename
            button.text = "Download"
            button.textSize = 11.0f

            val paramsB = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT
            )
            paramsB.gravity = Gravity.END
            button.layoutParams = paramsB
            button.setBackgroundResource(R.drawable.button_custom)

//            val exactPath = File("$path/${button.tag}")
            if (exactPath.exists()) {
                button.isEnabled = false
            } else {
                button.setOnClickListener(View.OnClickListener {
                    Toast.makeText(context, "Clicked on button ${button.tag}", Toast.LENGTH_SHORT).show()
                    if (downloadFiles.reqPermission(requireActivity(), button.tag.toString(), radioButton)) {
                        button.isEnabled = false
                    }
                })
            }

            radioGroup.addView(radioButton)
            radioGroup.addView(textView)
            radioGroup.addView(button)
        }

        // default button (first button)
//        val defButton = view.findViewById<Button>(R.id.mobilenet_v1_224_tflite)
//        val exactPath = File("$path/${defButton.tag}")
//
//        if (exactPath.exists()) {   // the file is already downloaded
//            defButton.isEnabled = false
//        } else {  // the file is not downloaded
//            defButton.setOnClickListener(View.OnClickListener {
//                Toast.makeText(context, "Clicked on button ${defButton.tag}", Toast.LENGTH_LONG).show()
//                defButton.isEnabled = false
//                downloadFiles.reqPermission(requireActivity(), defButton.tag.toString())
//            })
//        }

        // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +
        // +           set INITIAL RadioGroup SELECTION            +
        // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +

        // get saved model Id from SharedPreferences
        val id = prefs.getInt("model", 0)
        var r = view.findViewById<RadioButton>(id)
        if (id == 0 || r == null) {
            r = view.findViewById(R.id.radioButton_QuantizedMobileNet)
        }

        // set initial selection
        r!!.isChecked = true


        // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +
        // +           NETWORK SELECTOR via RadioGroup             +
        // + + + + + + + + + + + + + + + + + + + + + + + + + + + + +

        radioGroup.setOnCheckedChangeListener { group, checkedId -> // perform haptic feedback
            group.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_PRESS)
            var modelId = 0
            // save selection to SharedPreferences
            prefEditor.run {
                // check if it is the default RadioButton
                if (checkedId != R.id.radioButton_QuantizedMobileNet) {
                    modelId = view.findViewById<View>(checkedId).id
                }
                putInt("model", modelId) // save selection to SharedPreferences
                apply()
            }
            Classifier.onConfigChanged() // trigger classifier update
        }
    }

//    override fun onRequestPermissionsResult(
//        requestCode: Int,
//        permissions: Array<out String>,
//        grantResults: IntArray
//    ) {
//        when (requestCode) {
//            downloadFiles.STORAGE_PERMISSION_CODE -> {
//                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
//                    downloadFiles.startDownloading()
//                }
//                else {
//                    Toast.makeText(context, "Permission Denied", Toast.LENGTH_LONG).show()
//                }
//            }
//        }
//    }

    companion object {
        fun newInstance(): ModelSelectorFragment {
            return ModelSelectorFragment()
        }
    }
}